pennylane/gradients/hadamard_gradient.py

```diff
@@ -309,218 +290,122 @@ def hadamard_grad(
-            for m in tape.measurements:
-                if m.obs:
-                    obs_new = [m.obs]
-                else:
-                    m_wires = m.wires if len(m.wires) > 0 else tape.wires
-                    obs_new = [qml.Z(i) for i in m_wires]
-
-                obs_new.append(qml.Y(aux_wire))
-                obs_new = qml.prod(*obs_new)
-
-                if isinstance(m, qml.measurements.ExpectationMP):
-                    measurements.append(qml.expval(op=obs_new))
-                else:
-                    measurements.append(qml.probs(op=obs_new))
-
-            new_tape = qml.tape.QuantumScript(ops=ops, measurements=measurements, shots=tape.shots)
-            _rotations, _measurements = qml.tape.tape.rotations_and_diagonal_measurements(new_tape)
-            new_ops = new_tape.operations + _rotations
-            new_tape = qml.tape.QuantumScript(
-                new_ops,
-                _measurements,
-                shots=new_tape.shots,
-                trainable_params=new_tape.trainable_params,
-            )
-
-            num_tape += 1
+            measurements = (_new_measurement(mp, aux_wire, tape.wires) for mp in tape.measurements)
+            new_tape = qml.tape.QuantumScript(ops, measurements, shots=tape.shots)
 
             g_tapes.append(new_tape)
 
-        gradient_data.append(num_tape)
+    return g_tapes, partial(
+        processing_fn, coeffs=coeffs, tape=tape, generators_per_parameter=generators_per_parameter
+    )
+
+
+def _new_measurement(mp, aux_wire, all_wires: qml.wires.Wires):
+    obs = mp.obs or qml.prod(*(qml.Z(w) for w in mp.wires or all_wires))
+    new_obs = qml.simplify(obs @ qml.Y(aux_wire))
+    return type(mp)(obs=new_obs)
 
-    multi_measurements = len(tape.measurements) > 1
-    multi_params = len(tape.trainable_params) > 1
+
+def _get_pauli_generators(trainable_op):
+    """From a trainable operation, extract the unitary generators and their coefficients.
+    Any operator with a generator is supported.
+    """
+    generator = trainable_op.generator()
+    if generator.pauli_rep is None:
+        mat = qml.matrix(generator, wire_order=generator.wires)
+        op = qml.pauli_decompose(mat, wire_order=generator.wires)
+        return op.terms()
+    pauli_rep = generator.pauli_rep
+    id_pw = qml.pauli.PauliWord({})
+    if id_pw in pauli_rep:
+        del pauli_rep[qml.pauli.PauliWord({})]  # remove identity term
+    sum_op = pauli_rep.operation()
+    return sum_op.terms()
+
+
+def _postprocess_probs(res, measurement, tape):
+    projector = np.array([1, -1])
+    projector = qml.math.convert_like(projector, res)
+
+    num_wires_probs = len(measurement.wires)
+    if num_wires_probs == 0:
+        num_wires_probs = tape.num_wires
+    res = qml.math.reshape(res, (2**num_wires_probs, 2))
+    return qml.math.tensordot(res, projector, axes=[[1], [0]])
+
+
+def processing_fn(results: qml.typing.ResultBatch, tape, coeffs, generators_per_parameter):
```

Comment:
Pulled out into its own function to make it more clear what additional variables we rely on and reduce the overall complexity of the `_expval_hadamard_grad`.